backend/typescript/models/base.ts

```
@@ -1,20 +1,508 @@
+            if ((type as T).code === code) {
+                return type as T;
+            }
+        }
+        return null;
+    }
+
+    constructor(public code: string,
+                public shortDecodeText: string) {
+    }
+
+}
+
+// RAMObject ..........................................................................................................
+
+//noinspection ReservedWordAsName
+
+/**
+ * IRAMObject defines the common attributes that all transactional documents in RAM will contain. Alongside,
+ * ICodeDecode, these two interfaces define all the models in the system.
+ */
+export interface IRAMObject {
+    _id: any;
+    id: string;
+    isNew: boolean;
+    _createdAt: Date;
+    _updatedAt: Date;
+    _lastUpdatedByPartyRef: IResourceRef<IParty>;
+    deleteInd: boolean;
+    resourceVersion: string;
+    save(fn?: (err: any, product: this, numAffected: number) => void): Promise<this>;
+    delete(): void;
+    enforcePermissions(templates: Permissions, enforcers: IPermissionEnforcer<any>[]): Promise<Permissions>;
+    getPermissions(): Promise<Permissions>;
+}
+
+// exists for type safety only, do not add function implementations here
+export abstract class RAMObject implements IRAMObject {
+
+    public id: string;
+    public isNew: boolean;
+
+    constructor(public _id: any,
+                public _createdAt: Date,
+                public _updatedAt: Date,
+                public _lastUpdatedByPartyRef: IResourceRef<IParty>, // IResourceRef Model<IParty Model>
+                public deleteInd: boolean,
+                public resourceVersion: string) {
+        this.id = _id ? _id.toString() : undefined;
+    }
+
+    public save(fn?: (err: any, product: this, numAffected: number) => void): Promise<this> {
+        return null;
+    }
+
+    public delete(): void {
+        return null;
+    }
+
+    public enforcePermissions(templates: Permissions, enforcers: IPermissionEnforcer<any>[]): Promise<Permissions> {
+        return null;
```

This should throw an exception